server/proxy/sockproxy/sockproxy.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sockproxy
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/tls"
+	"log"
+	"net"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// SockProxy can multiple multiple HTTP and GRPC connections in the same address and port.
+// It is able to multiplex connections with and without TLS.
+// SockProxy can only forward connections to one HTTP or one GRPC server.
+// If only one of those is available it will forward all connections directly.
+// If both are available it will try to identify if the connection is HTTP,
+// and fall back to GRPC if is not HTTP.
+// It doesn't support HTTP/2 with Prior Knowledge, since is not common in the wild, and the golang
+// webserver requires an special handler.
+// This simplifies the heuristics, because it just need to detect if is HTTP/1.1, if the connection
+// is upgraded after that it will be keep working because it was already forwarded to the right
+// backend.
+type SockProxy struct {
+	lg    *zap.Logger
+	root  net.Listener
+	donec chan struct{}
+
+	secure    bool // serve TLS
+	insecure  bool // serve insecure
+	tlsConfig *tls.Config
+	http      *muxListener
+	grpc      *muxListener
+}
+
+// NewSockProxy creates a new SockProxy.
+func New(lg *zap.Logger, listener net.Listener) *SockProxy {
+	if lg == nil {
+		lg = zap.NewNop()
+	}
+	return &SockProxy{
+		root:  listener,
+		lg:    lg,
+		donec: make(chan struct{}),
+	}
+}
+
+func (c *SockProxy) Insecure() {
```

Comment:
Configuring proxy via calling methods doesn't seem great. Maybe `New` can just accept `Config` struct with all options?

Reply:
yes, absolutely, this was a quick sketch to get feedback, I was thinking in using variadic functions or a config struct